src/algebra/lie/centralizer.lean

```diff
@@ -0,0 +1,180 @@
+/-
+Copyright (c) 2022 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.lie.abelian
+import algebra.lie.ideal_operations
+import algebra.lie.quotient
+
+/-!
+# The centralizer of a Lie submodule and the normalizer of a Lie subalgebra.
+
+Given a Lie module `M` over a Lie subalgebra `L`, the centralizer of a Lie submodule `N ⊆ M` is
+the Lie submodule with underlying set `{ m | ∀ (x : L), ⁅x, m⁆ ∈ N }`.
+
+The lattice of Lie submodules thus has two natural operations, the centralizer: `N ↦ N.centralizer`
+and the ideal operation: `N ↦ ⁅⊤, N⁆`; these are adjoint, i.e., they form a Galois connection. This
+adjointness is the reason that we may define nilpotency in terms of either the upper or lower
+central series.
+
+Given a Lie subalgebra `H ⊆ L`, we may regard `H` as a Lie submodule of `L` over `H`, and thus
+consider the centralizer. This turns out to be a Lie subalgebra and is known as the normalizer.
+
+## Main definitions
+
+  * `lie_submodule.centralizer`
+  * `lie_subalgebra.normalizer`
+  * `lie_submodule.gc_top_lie_centralizer`
+
+## Tags
+
+lie algebra, centralizer, normalizer
+-/
+
+variables {R L M : Type*}
+variables [comm_ring R] [lie_ring L] [lie_algebra R L]
+variables [add_comm_group M] [module R M] [lie_ring_module L M] [lie_module R L M]
+
+namespace lie_submodule
+
+variables (N : lie_submodule R L M) {N₁ N₂ : lie_submodule R L M}
+
+/-- The centralizer of a Lie submodule. -/
+def centralizer : lie_submodule R L M :=
+{ carrier   := { m | ∀ (x : L), ⁅x, m⁆ ∈ N },
+  add_mem'  := λ m₁ m₂ hm₁ hm₂ x, by {  rw lie_add, exact N.add_mem' (hm₁ x) (hm₂ x), },
+  zero_mem' := λ x, by simp,
+  smul_mem' := λ t m hm x, by { rw lie_smul, exact N.smul_mem' t (hm x), },
+  lie_mem   := λ x m hm y, by { rw leibniz_lie, exact N.add_mem' (hm ⁅y, x⁆) (N.lie_mem (hm y)), } }
+
+@[simp] lemma mem_centralizer (m : M) :
+  m ∈ N.centralizer ↔ ∀ (x : L), ⁅x, m⁆ ∈ N :=
+iff.rfl
+
+lemma le_centralizer : N ≤ N.centralizer :=
+begin
+  intros m hm,
+  rw mem_centralizer,
+  exact λ x, N.lie_mem hm,
+end
+
+lemma centralizer_inf :
+   (N₁ ⊓ N₂).centralizer = N₁.centralizer ⊓ N₂.centralizer :=
+by { ext, simp [← forall_and_distrib], }
+
+@[mono] lemma centalizer_mono (h : N₁ ≤ N₂) :
```

Comment:
I am not very familiar with the `@[mono]` attribute: does it produce automatically the fact that `centralizer` is monotonic?

Reply:
`@[mono]` is a decorator that makes lemmas accessible to the `mono` tactic, somewhat similar to `simp`.

Reply:
Sorry, I got confused by the name. Should the name mention `comap`?

Reply:
Erm, now I'm confused: there's no `comap` in either the name or the statement.

Reply:
Sorry, I don't know why I read the wrong line, forget my comment.